ga4gh/backend.py

```diff
@@ -327,6 +297,110 @@ def getReferenceSetByName(self, name):
             raise exceptions.ReferenceSetNameNotFoundException(name)
         return self._referenceSetNameMap[name]
 
+
+class EmptyDataRepository(AbstractDataRepository):
+    """
+    A data repository that contains no data
+    """
+    def __init__(self):
+        super(EmptyDataRepository, self).__init__()
+
+
+class SimulatedDataRepository(AbstractDataRepository):
+    """
+    A data repository that is simulated
+    """
+    def __init__(
+            self, randomSeed=0, numDatasets=2,
+            numVariantSets=1, numCalls=1, variantDensity=0.5,
+            numReferenceSets=1, numReferencesPerReferenceSet=1,
+            numReadGroupSets=1, numReadGroupsPerReadGroupSet=1,
+            numAlignments=2):
+        super(SimulatedDataRepository, self).__init__()
+
+        # References
+        for i in range(numReferenceSets):
+            localId = "referenceSet{}".format(i)
+            seed = randomSeed + i
+            referenceSet = references.SimulatedReferenceSet(
+                localId, seed, numReferencesPerReferenceSet)
+            self.addReferenceSet(referenceSet)
+
+        # Datasets
+        for i in range(numDatasets):
+            seed = randomSeed + i
+            localId = "simulatedDataset{}".format(i)
+            referenceSet = self.getReferenceSetByIndex(i % numReferenceSets)
+            dataset = datasets.SimulatedDataset(
+                localId, referenceSet=referenceSet, randomSeed=seed,
+                numCalls=numCalls, variantDensity=variantDensity,
+                numVariantSets=numVariantSets,
+                numReadGroupSets=numReadGroupSets,
+                numReadGroupsPerReadGroupSet=numReadGroupsPerReadGroupSet,
+                numAlignments=numAlignments)
+            self.addDataset(dataset)
+
+
+class FileSystemDataRepository(AbstractDataRepository):
+    """
+    A data repository based on the file system
+    """
+    def __init__(self, dataDir):
+        super(FileSystemDataRepository, self).__init__()
+        self._dataDir = dataDir
+        sourceDirNames = ["referenceSets", "datasets"]
+        constructors = [
+            references.HtslibReferenceSet, datasets.FileSystemDataset]
+        objectAdders = [self.addReferenceSet, self.addDataset]
+        for sourceDirName, constructor, objectAdder in zip(
+                sourceDirNames, constructors, objectAdders):
+            sourceDir = os.path.join(self._dataDir, sourceDirName)
+            for setName in os.listdir(sourceDir):
+                relativePath = os.path.join(sourceDir, setName)
+                if os.path.isdir(relativePath):
+                    objectAdder(constructor(setName, relativePath, self))
+
+
+class AbstractBackend(object):
```

Comment:
One question here: do we need the Backend hierarchy any more? We could just have a single class now, ``Backend``, which takes a ``DataRepository`` instance as its argument. We can then get rid of ``SimulatedBackend``, ``FileSystemBackend`` and ``EmptyBackend``, which are all a bit pointless now. I think the only thing we'd need to update based on this is the frontend.configure() function, which would need a very slight refactor.

What do you think?